### PR TITLE
Rework simpler conditional offscreen for screen readback support.

### PR DESCRIPTION
**Pull request by flar**
This solution for the conditional offscreen allows us to check for layers that read from the surface in Preroll instead of in the constructors. There are no longer any flags propagating to parents or any other potentially quadratic recursions compared to the previous solution.

I have an FML_LOG(INFO) in the code that lets you check on the status of when we use an offscreen by running an app with `flutter run --verbose-system-logs`

Things TBD:
- run more tests (apps and examples to make sure they render correctly)

Former TBD things completed:
- specifically run the [BackdropFilter devicelab benchmark](https://github.com/flutter/flutter/blob/master/dev/benchmarks/macrobenchmarks/lib/src/backdrop_filter.dart) before/after to verify no loss of performance
- specifically run the LinearProgressApp and check its energy usage in Xcode and CPU/GPU usage with [the gauge in flutter-packages](https://github.com/flutter/packages/tree/master/packages/gauge)
- write (and then run) unit tests
- remove offscreen support from gpu_surface_gl (currently it is just bypassed, but it can be deleted)
- my Fuchsia build seems to be broken now so I can only marginally claim that it still compiles

### CODE REVIEW

**Comment by flar** [comment]
There is boilerplate in a number of layers that use savelayer to save the flag, preroll on children, and restore it. I could either provide PrerollWithSavelayer() and PrerollChildrenWithSavelayer() methods/inlines in ContainerLayer, or modify the existing Preroll()/PrerollChildren() methods to conditionally provide this service - thoughts?

File: shell/gpu/gpu_surface_gl.cc [hunk 1]
```diff
@@ -285,7 +285,8 @@ std::unique_ptr<SurfaceFrame> GPUSurfaceGL::AcquireFrame(const SkISize& size) {
         return weak ? weak->PresentSurface(canvas) : false;
       };
 
-  return std::make_unique<SurfaceFrame>(surface, submit_callback);
+  return std::make_unique<SurfaceFrame>(
+      surface, !delegate_->UseOffscreenSurface(), submit_callback);
```
Comment (flar) [inline shell/gpu/gpu_surface_gl.cc]: delegate->UseOffscreenSurface should probably be renamed given its new purpose. This can be done when I remove all of the offscreen logic from gpu_surface_gl.

File: shell/gpu/gpu_surface_metal.mm [hunk 2]
```diff
@@ -152,7 +152,7 @@
     return true;
   };
 
-  return std::make_unique<SurfaceFrame>(std::move(surface), submit_callback);
+  return std::make_unique<SurfaceFrame>(std::move(surface), true, submit_callback);
```
Comment (flar) [inline shell/gpu/gpu_surface_metal.mm]: Need to verify that we are using the right flags with the metal surface creation to support readback. Finding the optimal settings for Metal will be work for another PR, but we need to make sure the boolean used here is correct for now.

File: shell/gpu/gpu_surface_vulkan.cc [hunk 3]
```diff
@@ -40,7 +40,7 @@ std::unique_ptr<SurfaceFrame> GPUSurfaceVulkan::AcquireFrame(
     }
     return weak_this->window_.SwapBuffers();
   };
-  return std::make_unique<SurfaceFrame>(std::move(surface),
+  return std::make_unique<SurfaceFrame>(std::move(surface), true,
```
Comment (flar) [inline shell/gpu/gpu_surface_vulkan.cc]: Need to verify that this is the appropriate value for Vulkan surfaces.

File: shell/platform/fuchsia/flutter/surface.cc [hunk 4]
```diff
@@ -26,8 +26,10 @@ bool Surface::IsValid() {
 std::unique_ptr<flutter::SurfaceFrame> Surface::AcquireFrame(
     const SkISize& size) {
   return std::make_unique<flutter::SurfaceFrame>(
-      nullptr, [](const flutter::SurfaceFrame& surface_frame,
-                  SkCanvas* canvas) { return true; });
+      nullptr, true,
```
Comment (flar) [inline shell/platform/fuchsia/flutter/surface.cc]: Need to verify if this is the right value for surface_supports_readback for Fuchsia surfaces.

**Comment by flar** [comment]
We will need a further PR to investigate the best way to handle Metal surfaces. For this particular PR, I only want to focus on compatibility with what is working today and to enable the optimization for GL surfaces. I will file an Issue when this PR lands to investigate the Metal possibilities.

To be clear, Metal has a mutable property indicating whether a surface is only used for rendering, or whether it can be used for other operations - including the readback we need for surface_supports_readback. As a result, we may want to upgrade this fix to a 2-way communication protocol that allows us to inform the Metal surface when the tree is in a state that requires readback and let it update its settings. For now, with this fix, Metal will have to choose one way and then set the appropriate value for whether the properties it chose allows readback or not and that decision will be static.

For this PR, I'm making the assumption that the Metal properties were already requesting readback capabilities from the platform and I've set the flag to indicate that we can readback from a Metal surface. I want to verify that assumption, but if that isn't true then it means that our current Metal support has a bug that prevents BackdropFilters from working, so I doubt the assumption is incorrect.

Tagging @chinmaygarde for specific review in this area.

File: flow/layers/layer_tree.h [hunk 5]
```diff
@@ -22,7 +22,7 @@ class LayerTree {
  public:
   LayerTree();
 
-  void Preroll(CompositorContext::ScopedFrame& frame,
+  bool Preroll(CompositorContext::ScopedFrame& frame,
```
Comment (liyuqian) [inline flow/layers/layer_tree.h]: This public method needs documentation. The doc would be very useful especially because now it also a return.

File: shell/gpu/gpu_surface_gl.cc [hunk 6]
```diff
@@ -228,7 +228,7 @@ bool GPUSurfaceGL::CreateOrUpdateSurfaces(const SkISize& size) {
     return false;
   }
 
-  if (delegate_->UseOffscreenSurface()) {
+  if (false && delegate_->UseOffscreenSurface()) {
```
Comment (liyuqian) [inline shell/gpu/gpu_surface_gl.cc]: This seems to be a debugging hack. Maybe put a `[WIP]` in the PR title to notify that this PR includes work in progress.

File: shell/gpu/gpu_surface_metal.mm (hunk 2 above)
Comment (flar) [reply shell/gpu/gpu_surface_metal.mm]: Confirmed. Line 28 in the constructor in this same file specifically sets the "framebuffer only" flag on the CAMetalLayer to NO:

```
26:  // Flutter needs to read from the color attachment in cases where there are effects such as
27:  // backdrop filters.
28:  layer.get().framebufferOnly = NO;
```

File: flow/layers/backdrop_filter_layer.cc [hunk 7]
```diff
@@ -9,6 +9,17 @@ namespace flutter {
 BackdropFilterLayer::BackdropFilterLayer(sk_sp<SkImageFilter> filter)
     : filter_(std::move(filter)) {}
 
+void BackdropFilterLayer::Preroll(PrerollContext* context,
```
Comment (liyuqian) [inline flow/layers/backdrop_filter_layer.cc]: Why no let `Layer::Preroll` also return `bool` like the `LayerTree::Preroll` does? By returning bool, we can simplify most `Layer::Preroll` as

```
bool XyzLayer::Preroll(...) {
  bool children_needs_read_from_surface = ContainerLayer::Preroll(...);
  return this_layer_needs_read_from_surface() || !use_save_layer() && children_needs_read_from_surface;
}

bool ContainerLayer::Preroll(...) {
  bool children_needs_read_from_surface = false;
  for (child in children) {
    children_needs_read_from_surface ||= child->Preroll(...);
  }
  return children_needs_read_from_surface;
}
```

File: shell/gpu/gpu_surface_gl.cc (hunk 1 above)
Comment (flar) [reply shell/gpu/gpu_surface_gl.cc]: Done in the next commit. It will now be delegate->SurfaceSupportsReadback().

File: shell/gpu/gpu_surface_gl.cc (hunk 6 above)
Comment (flar) [reply shell/gpu/gpu_surface_gl.cc]: I mentioned this in the description. It is already done pending my next commit.

Comment (flar) [reply shell/gpu/gpu_surface_gl.cc]: I added the WIP label since I have some more changes coming, but I do want a review of the current approach before I get too far in finalizing the tests and such.

File: flow/layers/backdrop_filter_layer.cc (hunk 7 above)
Comment (flar) [reply flow/layers/backdrop_filter_layer.cc]: Eventually Preroll will collect more information, such as the dirty regions. Even now the context is tracking multiple pieces of information that are modified and overridden on a layer by layer basis and so it seems that is the best way to collect any data that is negotiated between layers in the tree.

File: flow/layers/layer_tree.h (hunk 5 above)
Comment (flar) [reply flow/layers/layer_tree.h]: I will add a comment. Eventually this return value will need to grow to more than just a boolean as more data will be collected from the tree. I decided against creating a return structure just yet until that future work (well outside the scope of this PR) is completed.

File: shell/gpu/gpu_surface_vulkan.cc (hunk 3 above)
Comment (flar) [reply shell/gpu/gpu_surface_vulkan.cc]: This appears to be correct as we specify no clipping when constructing the swapchain in vulkan_swapchain.cc:

```
133:      .clipped = VK_FALSE,
```

According to Vulkan documentation, clipped=TRUE allows them to drop pixels for parts of the surface that are hidden behind other windows or otherwise clipped out and the application should set this to FALSE if they expect to read back the pixels.

File: shell/platform/fuchsia/flutter/surface.cc (hunk 4 above)
Comment (flar) [reply shell/platform/fuchsia/flutter/surface.cc]: Trivially, this routine is returning a dummy SurfaceFrame (note that the surface itself is nullptr) so the value doesn't really matter.

File: flow/layers/layer_tree.h (hunk 5 above)
Comment (flar) [reply flow/layers/layer_tree.h]: Comment is available in the next commit.

**Comment by flar** [comment]
backdrop_filter_perf.dart benchmark results:
```
                             before fix      after fix
                             ----------      ---------

avg build time run 1:         0.692 ms        0.690 ms
avg build time run 2:         0.672 ms        0.684 ms
avg build time run 3:         0.711 ms        0.697 ms
overall avg build time:       0.6917          0.6903    (<.2% reduction)

avg raster time run 1:        145.8 ms        146.8 ms
avg raster time run 2:        145.7 ms        148.2 ms
avg raster time run 3:        145.7 ms        147.7 ms
overall avg raster time:      145.7           147.6     (1.25% increase)
```
Running the LinearProgressApp:
```
                         before fix      after fix
                         ----------      ---------
Xcode energy rating:     Very High          Low
(within/for)              < 40sec         > 10min

GPU gauge run 1:           16.8%           8.8%
GPU gauge run 2:           15.2%           8.4%
GPU gauge run 3:           15.4%           8.6%
GPU gauge run 4:           15.0%           8.4%
GPU gauge average:         15.6%           8.6%    (45% reduction)

CPU gauge run 1:           19.0%           16.0%
CPU gauge run 2:           19.6%           15.4%
CPU gauge run 3:           19.8%           16.9%
CPU gauge run 4:           18.8%           17.3%
CPU gauge average:         19.3%           16.4%   (15% reduction)

Combined GPU/CPU                                   (53% reduction)

(Real-life drain measurements will be updated when I get a spare 4 hours)
Real-life battery drain:    40%             TBD%    (TBD% reduction
Compared to blank screen:   16%             TBD%    (TBD% reduction)

Source for the app:

    return MaterialApp(
      title: 'Demo',
      home: Scaffold(
        body: Center(
          child: LinearProgressIndicator(),
        ),
      ),
    );
```

**Comment by flar** [comment]
It is outside the scope of this PR, but it seems like all of the variants of ClipFooLayer could be rolled into a common parent class with extremely minor implementations for their differences (things like returning the clip bounds and installing the clip shape into a mutator stack and/or a canvas, everything else should be identical among them).

**Comment by flar** [comment]
I've removed the WIP label as this is pretty much complete and all that is happening now is running various apps for testing - and waiting for a 4 hour time slot when I can run the "real world battery rundown" tests.

File: flow/layers/clip_path_layer_unittests.cc [hunk 8]
```diff
@@ -192,5 +192,69 @@ TEST_F(ClipPathLayerTest, PartiallyContainedChild) {
            MockCanvas::DrawCall{1, MockCanvas::RestoreData{0}}}));
 }
```
Comment (liyuqian) [inline flow/layers/clip_path_layer_unittests.cc]: See comments on [physical_shape_layer_unittests.cc](https://github.com/flutter/engine/pull/14104/files#diff-0c5e7bba6ccd82eea4c050757fd2fb91) about avoidng macros.

File: flow/layers/clip_rect_layer_unittests.cc [hunk 9]
```diff
@@ -190,5 +190,68 @@ TEST_F(ClipRectLayerTest, PartiallyContainedChild) {
            MockCanvas::DrawCall{1, MockCanvas::RestoreData{0}}}));
 }
 
+#define CHECK_READBACK(layer, before, after)                       \
```
Comment (liyuqian) [inline flow/layers/clip_rect_layer_unittests.cc]: See comments on [physical_shape_layer_unittests.cc](https://github.com/flutter/engine/pull/14104/files#diff-0c5e7bba6ccd82eea4c050757fd2fb91) about avoidng macros.

File: flow/layers/clip_rrect_layer_unittests.cc [hunk 10]
```diff
@@ -195,5 +195,69 @@ TEST_F(ClipRRectLayerTest, PartiallyContainedChild) {
            MockCanvas::DrawCall{1, MockCanvas::RestoreData{0}}}));
 }
 
+#define CHECK_READBACK(layer, before, after)                       \
```
Comment (liyuqian) [inline flow/layers/clip_rrect_layer_unittests.cc]: See comments on [physical_shape_layer_unittests.cc](https://github.com/flutter/engine/pull/14104/files#diff-0c5e7bba6ccd82eea4c050757fd2fb91) about avoidng macros.

File: flow/compositor_context.cc [hunk 11]
```diff
@@ -36,10 +36,11 @@ std::unique_ptr<CompositorContext::ScopedFrame> CompositorContext::AcquireFrame(
     ExternalViewEmbedder* view_embedder,
     const SkMatrix& root_surface_transformation,
     bool instrumentation_enabled,
+    bool surface_supports_readback,
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/compositor_context.cc]: This is a property on the surface we'll be rendering onto. Surfaces support or don't support readback, they don't "need readback". The "needs" property is something you get from the layer(tree) later in the process (see below).

[the remark this answers is not in the document]
Comment (liyuqian) [reply flow/compositor_context.cc]: Oh, you're right. I mistake it with the `tree_needs_readback`. Resolving this.

File: flow/layers/physical_shape_layer_unittests.cc [hunk 12]
```diff
@@ -202,5 +202,74 @@ TEST_F(PhysicalShapeLayerTest, ElevationComplex) {
                0, MockCanvas::DrawPathData{layer_path, layer_paint}}}));
 }
 
+#define MAKE_PHYSICAL_SHAPE_LAYER(clip)                              \
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc]: For that one, that is reasonable, but for the others, they refer to info that is not available in a static method and so I made them macros to avoid having to pass in a huge argument list:

Here is the context you can't fetch on your own
And here is the SkMatrix I already created for you (arguably could be created every time in the called method)
And here is the path I made for the clip (again, could be created every time)
And here is the (mock) child I made which may or may not do readback.
Etc.

Some of them could just be created locally on every call, though...

File: flow/layers/physical_shape_layer_unittests.cc [hunk 13]
```diff
@@ -202,5 +202,74 @@ TEST_F(PhysicalShapeLayerTest, ElevationComplex) {
...
+  std::make_shared<PhysicalShapeLayer>(SK_ColorGREEN, SK_ColorBLACK, \
+                                       0.0f, /* elevation */         \
+                                       layer_path, clip)
+
+#define CHECK_READBACK(layer, before, after)                       \
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc]: (I get it, I know where all the macros are. I closed all of those comments in favor of a single one here, but the caveats I mentioned above apply for most of them. The readability of the code gets really strained unless I can find ways to simplify the data being passed down...)

File: flow/layers/layer.h [hunk 14]
```diff
@@ -49,6 +49,7 @@ struct PrerollContext {
   MutatorsStack& mutators_stack;
   SkColorSpace* dst_color_space;
   SkRect cull_rect;
+  bool layer_reads_from_surface;
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/layer.h]: This could use a better name, but I don't like "skia layer needs readback" because skia doesn't have needs, they provide support. I'll think of something better that reflects that this is the current status of whether or not we've encountered an engine layer that uses a read-back operation...

"readback_operation_detected"?

Still thinking...

File: flow/layers/physical_shape_layer_unittests.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc]: (And, I did originally use methods, but the list of arguments were less readable than using macros - I'll try again and see if we can have the best of both worlds - maybe if I don't make them static...)

[the remark this answers is not in the document]
Comment (liyuqian) [reply flow/layers/physical_shape_layer_unittests.cc]: I think you can create a sublass `class PhysicalShapeLayerTest : public LayerTest` and save whatever you want to pass in the member fields of `PhysicalShapeLayerTest`.

[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc]: It was easy enough to just create new objects for each test case and the test still runs in 0ms. The only thing needed to be passed in is the PrerollContext.

File: flow/layers/physical_shape_layer_unittests.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/physical_shape_layer_unittests.cc]: Done.

File: flow/layers/layer.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/layer.h]: I renamed it to subtree_performs_readback_operation and created an AutoPrerollSaveLayer class to manage the flags for saveLayer containers.

[the remark this answers is not in the document]
Comment (liyuqian) [reply flow/layers/layer.h]: Hmm... it's still unclear to me what's the root of the subtree. At a first glance, I'd think the root to be the current layer that's doing preroll. But it's possible that the subtree of the current layer doesn't do readback, back the `subtree_performs_readback_operation` is still true because its older sibling performs readback.

How about `needs_readback_from_current_skia_layer`? Then everything seems to be more intuitive: whenever Skia saveLayer is called, we're creating a new Skia layer, so we clear this flag.

File: flow/layers/clip_path_layer.cc [hunk 15]
```diff
@@ -22,7 +22,10 @@ void ClipPathLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
   SkRect clip_path_bounds = clip_path_.getBounds();
   children_inside_clip_ = context->cull_rect.intersect(clip_path_bounds);
   if (children_inside_clip_) {
+    Layer::AutoPrerollSaveLayer save =
+        Layer::AutoPrerollSaveLayer::Create(context, uses_save_layer());
     context->mutators_stack.PushClipPath(clip_path_);
+
```
Comment (liyuqian) [inline flow/layers/clip_path_layer.cc]: Nit: extra new line.

File: flow/layers/clip_path_layer.h [hunk 16]
```diff
@@ -17,6 +17,10 @@ class ClipPathLayer : public ContainerLayer {
 
   void Paint(PaintContext& context) const override;
 
+  bool uses_save_layer() const {
```
Comment (liyuqian) [inline flow/layers/clip_path_layer.h]: Nit: not a simple getter so the naming style is `UsesSaveLayer`.

File: flow/layers/clip_rrect_layer_unittests.cc [hunk 17]
```diff
@@ -195,5 +195,65 @@ TEST_F(ClipRRectLayerTest, PartiallyContainedChild) {
            MockCanvas::DrawCall{1, MockCanvas::RestoreData{0}}}));
 }
 
+static bool ReadbackResult(PrerollContext* context,
```
Comment (liyuqian) [inline flow/layers/clip_rrect_layer_unittests.cc]: Extra extra nit: I felt that you can write template functions in a header file like `clip_layer_readback_unittests.h` to merge similar code in clip_rect, clip_rrect, and clip_path unit tests. This is of course not very important and it shouldn't block you from merging this PR. Feel free to come back to play with it later when you have time and mood :)

**Comment by liyuqian** [comment]
BTW, before merging, please make sure to put https://github.com/flutter/flutter/issues/31865 in the PR/commit description for future reference.

File: flow/layers/layer.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/layer.h]: The root of the subtree is basically since the last saveLayer in this case. It's really a property for the current drawable surface. How about "readback_detected_on_current_drawable" or something?

You keep suggesting names containing "skia", but I feel that the name of the graphics library is an implementation detail. It's really that the "thing" we are rendering onto is a drawable or surface and the fact that we get them from Skia is incidental to the property being tracked.

File: flow/layers/backdrop_filter_layer.cc [hunk 18]
```diff
@@ -9,6 +9,13 @@ namespace flutter {
...
+                                  const SkMatrix& matrix) {
+  Layer::AutoPrerollSaveLayer save =
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/backdrop_filter_layer.cc]: We currently have AutoSaveLayer which is a contextual object for wrapping a tree descent with code that needs to be done to set up and take down the save layer. It actually does more than that, but everything it does is related to managing a save layer.

The new similarly tasked object I added is doing the same thing during Preroll, it is managing the setup and take down of state that needs to be managed for a save layer during the Preroll operation.  Perhaps having both names match more would help?

AutoPaintSaveLayer
AutoPrerollSaveLayer

AutoSaveLayer::Paint
AutoSaveLayer::Preroll

?

File: flow/layers/clip_rrect_layer_unittests.cc (hunk 17 above)
Comment (flar) [reply flow/layers/clip_rrect_layer_unittests.cc]: On a related note, I pointed out elsewhere that we could probably have all of the ClipFooLayer implementations subclass off of a ClipLayerBase class and all they'd need to implement would be 2 or 3 very short helper methods.

File: flow/layers/layer.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/layer.h]: => surface_needs_readback

File: flow/layers/backdrop_filter_layer.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/backdrop_filter_layer.cc]: I went with AutoPrerollSaveLayerState - we discussed ManagePrerollSaveLayerState, but I felt that Auto reflected its scoped actions. Manage sounds like a 1-time operation (and was creating awkward line lengths).

**Comment by flar** [comment]
This change should help with https://github.com/flutter/flutter/issues/31865

In the best case, with only a single animating LinearProgressIndicator on the screen the improvement can be as much as a 50% reduction in combined CPU/GPU usage, but larger apps should see less of that reduction as the cost of the extra offscreen layer will gradually be dominated by other work the application is doing.

Cupertino apps will automatically have a BackdropFilter in a few places unless they take steps to disable them. For example, the header will have a translucent background color by default and a blur filter is applied under it. If you set an opaque background color then the backdrop will not be applied.
